File: novascript/src/frontend/lexer.ts

// src/frontend/lexer.ts

export enum TokenType {
  // Literals
  Number,
  String,
  Identifier,
  
  // Keywords
  Let,
  Const,
  Global,
  Fn,
  If,
  Else,
  While,
  For,
  From,
  To,
  Return,
  Include,
  And,
  Or,
  Not,
  Is,
  Isnt,
  
  // Symbols
  Assign,      // =
  PlusEquals,  // +=
  MinusEquals, // -=
  Equals,      // == or is
  NotEquals,   // != or isnt
  LessThan,    // <
  LessEquals,  // <=
  GreaterThan, // >
  GreaterEquals,// >=
  Plus,        // +
  Minus,       // -
  Times,       // *
  Power,       // **
  Slash,       // /
  Percent,     // %
  Ampersand,   // &
  Pipe,        // |
  Caret,       // ^
  ShiftLeft,   // <<
  ShiftRight,  // >>
  AndLogic,    // && or and
  OrLogic,     // || or or
  NotLogic,    // ! or not
  
  OpenParen,   // (
  CloseParen,  // )
  OpenBrace,   // {
  CloseBrace,  // }
  OpenBracket, // [
  CloseBracket,// ]
  Comma,
  Dot,
  Colon,       // :
  SemiColon,   // ;
  Arrow,       // =>
  DotDotDot,   // ...
  Question,    // ?
  NullCoalesce,// ??
  OptionalChain,// ?.

  // Keywords
  Switch,
  Case,
  Default,
  Try,
  Catch,
  Finally,
  Throw,
  Async,
  Await,
  Break,
  Continue,
  Export,
  
  // Special
  EOF,
}

const KEYWORDS: Record<string, TokenType> = {
  "let": TokenType.Let,
  "const": TokenType.Const,
  "global": TokenType.Global,
  "fn": TokenType.Fn,
  "if": TokenType.If,
  "else": TokenType.Else,
  "while": TokenType.While,
  "for": TokenType.For,
  "from": TokenType.From,
  "to": TokenType.To,
  "return": TokenType.Return,
  "include": TokenType.Include,
  "import": TokenType.Include,
  
  // Aliases (Mapped to logic tokens directly)
  "and": TokenType.AndLogic,
  "or": TokenType.OrLogic,
  "not": TokenType.NotLogic,
  "is": TokenType.Equals,
  "isnt": TokenType.NotEquals,

  "switch": TokenType.Switch,
  "case": TokenType.Case,
  "default": TokenType.Default,
  "try": TokenType.Try,
  "catch": TokenType.Catch,
  "finally": TokenType.Finally,
  "throw": TokenType.Throw,
  "async": TokenType.Async,
  "await": TokenType.Await,
  "break": TokenType.Break,
  "continue": TokenType.Continue,
  "export": TokenType.Export,
};

export interface Token {
  value: string;
  type: TokenType;
  line: number;
  column: number;
}

export function tokenize(sourceCode: string): Token[] {
  const tokens: Token[] = [];
  const src = sourceCode.split("");
  let line = 1;
  let column = 1;

  const pushToken = (value: string, type: TokenType) => {
      tokens.push({ value, type, line, column: column - value.length });
  };

  while (src.length > 0) {
    // 1. Handle Single-character Symbols
    if (src[0] === "(") {
      pushToken(src.shift()!, TokenType.OpenParen);
    } else if (src[0] === ")") {
      pushToken(src.shift()!, TokenType.CloseParen);
    } else if (src[0] === "{") {
      pushToken(src.shift()!, TokenType.OpenBrace);
    } else if (src[0] === "}") {
      pushToken(src.shift()!, TokenType.CloseBrace);
    } else if (src[0] === "[") {
      pushToken(src.shift()!, TokenType.OpenBracket);
    } else if (src[0] === "]") {
      pushToken(src.shift()!, TokenType.CloseBracket);
    } else if (src[0] === ",") {
      pushToken(src.shift()!, TokenType.Comma);
    } else if (src[0] === ":") {
      pushToken(src.shift()!, TokenType.Colon);
    } else if (src[0] === ";") {
      pushToken(src.shift()!, TokenType.SemiColon);
    }
 else if (src[0] === ".") {
      pushToken(src.shift()!, TokenType.Dot);
    } else if (src[0] === "+") {
        if (src[1] === "=") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "+=", type: TokenType.PlusEquals, line, column: column - 2 });
             continue;
        } else {
             pushToken(src.shift()!, TokenType.Plus);
        }
    } else if (src[0] === "-") {
        if (src[1] === "=") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "-=", type: TokenType.MinusEquals, line, column: column - 2 });
             continue;
        } else {
             pushToken(src.shift()!, TokenType.Minus);
        }
    } else if (src[0] === "*") {
        if (src[1] === "*") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "**", type: TokenType.Power, line, column: column - 2 });
             continue;
        } else {
             pushToken(src.shift()!, TokenType.Times);
        }
    } else if (src[0] === "/") {
      pushToken(src.shift()!, TokenType.Slash);
    } else if (src[0] === "%") {
      pushToken(src.shift()!, TokenType.Percent);
    } else if (src[0] === "=") {
        if (src[1] === "=") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "==", type: TokenType.Equals, line, column: column - 2 });
             continue;
        } else if (src[1] === ">") {
            src.shift(); src.shift(); column += 2;
            tokens.push({ value: "=>", type: TokenType.Arrow, line, column: column - 2 });
            continue;
        } else {
             pushToken(src.shift()!, TokenType.Assign);
        }
    }
 else if (src[0] === "!") {
        if (src[1] === "=") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "!=", type: TokenType.NotEquals, line, column: column - 2 });
             continue;
        } else {
            pushToken(src.shift()!, TokenType.NotLogic);
        }
    } else if (src[0] === "<") {
        if (src[1] === "=") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "<=", type: TokenType.LessEquals, line, column: column - 2 });
             continue;
        } else if (src[1] === "<") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "<<", type: TokenType.ShiftLeft, line, column: column - 2 });
             continue;
        } else {
             pushToken(src.shift()!, TokenType.LessThan);
        }
    } else if (src[0] === ">") {
        if (src[1] === "=") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: ">=", type: TokenType.GreaterEquals, line, column: column - 2 });
             continue;
        } else if (src[1] === ">") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: ">>", type: TokenType.ShiftRight, line, column: column - 2 });
             continue;
        } else {
             pushToken(src.shift()!, TokenType.GreaterThan);
        }
    } else if (src[0] === "&") {
        if (src[1] === "&") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "&&", type: TokenType.AndLogic, line, column: column - 2 });
             continue;
        } else {
             pushToken(src.shift()!, TokenType.Ampersand);
        }
    } else if (src[0] === "|") {
        if (src[1] === "|") {
             src.shift(); src.shift(); column += 2;
             tokens.push({ value: "||", type: TokenType.OrLogic, line, column: column - 2 });
             continue;
        } else {
             pushToken(src.shift()!, TokenType.Pipe);
        }
    } else if (src[0] === "^") {
        pushToken(src.shift()!, TokenType.Caret);
    } else if (src[0] === "?") {
        if (src[1] === "?") {
            src.shift(); src.shift(); column += 2;
            tokens.push({ value: "??", type: TokenType.NullCoalesce, line, column: column - 2 });
            continue;
        } else if (src[1] === ".") {
            src.shift(); src.shift(); column += 2;
            tokens.push({ value: "?.", type: TokenType.OptionalChain, line, column: column - 2 });
            continue;
        } else {
            pushToken(src.shift()!, TokenType.Question);
        }
    }

    // 2. Handle Comments, Whitespace, and Multi-character Tokens
    else if (src[0] === "#") {
        while(src.length > 0 && (src[0] as string) !== "\n") {
            src.shift();
        }
    } 
 else if (src[0] === "\n") {
        line++;
        column = 1;
        src.shift();
        continue;
    } else if (/\s/.test(src[0])) {
      src.shift();
      column++;
      continue;
    } else if (/[0-9]/.test(src[0])) {
        let num = "";
        while (src.length > 0 && /[0-9.]/.test(src[0])) {
            num += src.shift();
            column++;
        }
        tokens.push({ value: num, type: TokenType.Number, line, column: column - num.length });
        continue;
    } else if (src[0] === '"' || src[0] === "'") {
        const quote = src.shift()!;
        const startLine = line;
        const startColumn = column;
        column++;
        let str = "";
        while (src.length > 0 && (src[0] as string) !== quote) {
            if ((src[0] as string) === '\\') {
                src.shift(); column++;
                const escapeChar = src.shift();
                column++;
                if (!escapeChar) break;
                switch (escapeChar) {
                    case 'n': str += '\n'; break;
                    case 'r': str += '\r'; break;
                    case 't': str += '\t'; break;
                    case 'b': str += '\b'; break;
                    case 'f': str += '\f'; break;
                    case 'v': str += '\v'; break;
                    case '\\': str += '\\'; break;
                    case '"': str += '"'; break;
                    case "'": str += "'"; break;
                    case 'e': str += '\x1b'; break;
                    case 'x': {
                        let hex = (src.shift() || "") + (src.shift() || "");
                        column += 2;
                        str += String.fromCharCode(parseInt(hex, 16) || 0);
                        break;
                    }
                    default: str += escapeChar; break;
                }
            } else {
                str += src.shift();
                column++;
            }
        }
        if (src.length > 0 && src[0] === quote) {
            src.shift(); column++;
        }
        tokens.push({ value: str, type: TokenType.String, line: startLine, column: startColumn });
        continue;
    }
 else if (/[a-zA-Z_]/.test(src[0])) {
        let ident = "";
        while (src.length > 0 && /[a-zA-Z0-9_]/.test(src[0])) {
            ident += src.shift();
            column++;
        }
        
        const lowerIdent = ident.toLowerCase();
        const reserved = KEYWORDS[lowerIdent];
        if (reserved !== undefined) {
            tokens.push({ value: ident, type: reserved, line, column: column - ident.length });
        } else {
            tokens.push({ value: ident, type: TokenType.Identifier, line, column: column - ident.length });
        }
        continue;
    } else {
      console.error(`Unrecognized character found in source: ${src[0]} at line ${line}`);
      src.shift(); column++;
    }
    
    // Default increment for single char tokens handled by pushToken
    column++;
  }
  
  tokens.push({ type: TokenType.EOF, value: "EndOfFile", line, column });
  return tokens;
}

File: frontend/lexer.go
package frontend

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

type TokenType int

const (
	// Literals
	Number TokenType = iota
	StringToken
	IdentifierToken

	// Keywords
	Let
	Const
	Global
	Fn
	If
	Else
	While
	For
	From
	To
	Return
	Include
	And
	Or
	Not
	Is
	Isnt

	// Symbols
	Assign       // =
	PlusEquals   // +=
	MinusEquals  // -=
	Equals       // == or is
	NotEquals    // != or isnt
	LessThan     // <
	LessEquals   // <=
	GreaterThan  // >
	GreaterEquals// >=
	Plus         // +
	Minus        // -
	Times        // *
	Power        // **
	Slash        // /
	Percent      // %
	Ampersand    // &
	Pipe         // |
	Caret        // ^
	ShiftLeft    // <<
	ShiftRight   // >>
	AndLogic     // && or and
	OrLogic      // || or or
	NotLogic     // ! or not

	OpenParen    // (
	CloseParen   // )
	OpenBrace    // {
	CloseBrace   // }
	OpenBracket  // [
	CloseBracket // ]
	Comma
	Dot
	Colon        // :
	SemiColon    // ;
	Arrow        // =>
	DotDotDot    // ...
	Question     // ?
	NullCoalesce // ??
	OptionalChain// ?.

	// Keywords
	Switch
	Case
	Default
	Try
	Catch
	Finally
	Throw
	Async
	Await
	Break
	Continue
	Export

	// Special
	EOF
)

var KEYWORDS = map[string]TokenType{
	"let":      Let,
	"const":    Const,
	"global":   Global,
	"fn":       Fn,
	"if":       If,
	"else":     Else,
	"while":    While,
	"for":      For,
	"from":     From,
	"to":       To,
	"return":   Return,
	"include":  Include,
	"import":   Include,
	"and":      AndLogic,
	"or":       OrLogic,
	"not":      NotLogic,
	"is":       Equals,
	"isnt":     NotEquals,
	"switch":   Switch,
	"case":     Case,
	"default":  Default,
	"try":      Try,
	"catch":    Catch,
	"finally":  Finally,
	"throw":    Throw,
	"async":    Async,
	"await":    Await,
	"break":    Break,
	"continue": Continue,
	"export":   Export,
}

type Token struct {
	Value  string
	Type   TokenType
	Line   int
	Column int
}

func Tokenize(sourceCode string) []Token {
	var tokens []Token
	src := []rune(sourceCode)
	line := 1
	column := 1

	pushToken := func(value string, t TokenType) {
		tokens = append(tokens, Token{Value: value, Type: t, Line: line, Column: column - len(value)})
	}

	shift := func() string {
		c := string(src[0])
		src = src[1:]
		return c
	}

	for len(src) > 0 {
		c := src[0]
		if c == '(' {
			pushToken(shift(), OpenParen)
		} else if c == ')' {
			pushToken(shift(), CloseParen)
		} else if c == '{' {
			pushToken(shift(), OpenBrace)
		} else if c == '}' {
			pushToken(shift(), CloseBrace)
		} else if c == '[' {
			pushToken(shift(), OpenBracket)
		} else if c == ']' {
			pushToken(shift(), CloseBracket)
		} else if c == ',' {
			pushToken(shift(), Comma)
		} else if c == ':' {
			pushToken(shift(), Colon)
		} else if c == ';' {
			pushToken(shift(), SemiColon)
		} else if c == '.' {
			pushToken(shift(), Dot)
		} else if c == '+' {
			if len(src) > 1 && src[1] == '=' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "+=", Type: PlusEquals, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), Plus)
			}
		} else if c == '-' {
			if len(src) > 1 && src[1] == '=' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "-=", Type: MinusEquals, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), Minus)
			}
		} else if c == '*' {
			if len(src) > 1 && src[1] == '*' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "**", Type: Power, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), Times)
			}
		} else if c == '/' {
			pushToken(shift(), Slash)
		} else if c == '%' {
			pushToken(shift(), Percent)
		} else if c == '=' {
			if len(src) > 1 && src[1] == '=' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "==", Type: Equals, Line: line, Column: column - 2})
				continue
			} else if len(src) > 1 && src[1] == '>' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "=>", Type: Arrow, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), Assign)
			}
		} else if c == '!' {
			if len(src) > 1 && src[1] == '=' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "!=", Type: NotEquals, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), NotLogic)
			}
		} else if c == '<' {
			if len(src) > 1 && src[1] == '=' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "<=", Type: LessEquals, Line: line, Column: column - 2})
				continue
			} else if len(src) > 1 && src[1] == '<' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "<<", Type: ShiftLeft, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), LessThan)
			}
		} else if c == '>' {
			if len(src) > 1 && src[1] == '=' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: ">=", Type: GreaterEquals, Line: line, Column: column - 2})
				continue
			} else if len(src) > 1 && src[1] == '>' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: ">>", Type: ShiftRight, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), GreaterThan)
			}
		} else if c == '&' {
			if len(src) > 1 && src[1] == '&' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "&&", Type: AndLogic, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), Ampersand)
			}
		} else if c == '|' {
			if len(src) > 1 && src[1] == '|' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "||", Type: OrLogic, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), Pipe)
			}
		} else if c == '^' {
			pushToken(shift(), Caret)
		} else if c == '?' {
			if len(src) > 1 && src[1] == '?' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "??", Type: NullCoalesce, Line: line, Column: column - 2})
				continue
			} else if len(src) > 1 && src[1] == '.' {
				shift()
				shift()
				column += 2
				tokens = append(tokens, Token{Value: "?.", Type: OptionalChain, Line: line, Column: column - 2})
				continue
			} else {
				pushToken(shift(), Question)
			}
		} else if c == '#' {
			for len(src) > 0 && src[0] != '\n' {
				shift()
			}
		} else if c == '\n' {
			line++
			column = 1
			shift()
			continue
		} else if isWhitespace(c) {
			shift()
			column++
			continue
		} else if isDigit(c) {
			num := ""
			for len(src) > 0 && (isDigit(src[0]) || src[0] == '.') {
				num += shift()
				column++
			}
			tokens = append(tokens, Token{Value: num, Type: Number, Line: line, Column: column - len(num)})
			continue
		} else if c == '"' || c == '\'' {
			quote := shift()
			startLine := line
			startColumn := column
			column++
			str := ""
			for len(src) > 0 && string(src[0]) != quote {
				if src[0] == '\\' {
					shift()
					column++
					if len(src) == 0 {
						break
					}
					escapeChar := shift()
					column++
					switch escapeChar {
					case "n":
						str += "\n"
					case "r":
						str += "\r"
					case "t":
						str += "\t"
					case "b":
						str += "\b"
					case "f":
						str += "\f"
					case "v":
						str += "\v"
					case "\\":
						str += "\\"
					case "\"":
						str += "\""
					case "'":
						str += "'"
					case "e":
						str += "\x1b"
					case "x":
						if len(src) >= 2 {
							hex := shift() + shift()
							column += 2
							val, _ := strconv.ParseInt(hex, 16, 32)
							str += string(rune(val))
						}
					default:
						str += escapeChar
					}
				} else {
					str += shift()
					column++
				}
			}
			if len(src) > 0 && string(src[0]) == quote {
				shift()
				column++
			}
			tokens = append(tokens, Token{Value: str, Type: StringToken, Line: startLine, Column: startColumn})
			continue
		} else if isAlpha(c) {
			ident := ""
			for len(src) > 0 && isAlphaNumeric(src[0]) {
				ident += shift()
				column++
			}
			lowerIdent := strings.ToLower(ident)
			if reserved, ok := KEYWORDS[lowerIdent]; ok {
				tokens = append(tokens, Token{Value: ident, Type: reserved, Line: line, Column: column - len(ident)})
			} else {
				tokens = append(tokens, Token{Value: ident, Type: IdentifierToken, Line: line, Column: column - len(ident)})
			}
			continue
		} else {
			fmt.Printf("Unrecognized character found in source: %c at line %d\n", c, line)
			shift()
			column++
		}
		column++
	}

	tokens = append(tokens, Token{Type: EOF, Value: "EndOfFile", Line: line, Column: column})
	return tokens
}

func isWhitespace(r rune) bool {
	return r == ' ' || r == '\t' || r == '\r'
}

func isDigit(r rune) bool {
	return r >= '0' && r <= '9'
}

func isAlpha(r rune) bool {
	matched, _ := regexp.MatchString(`[a-zA-Z_]`, string(r))
	return matched
}

func isAlphaNumeric(r rune) bool {
	matched, _ := regexp.MatchString(`[a-zA-Z0-9_]`, string(r))
	return matched
}
